adapters/latest/src/main/java/org/apache/cassandra/sidecar/adapters/latest/CassandraAdapter.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.sidecar.adapters.trunk;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.driver.core.Metadata;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import org.apache.cassandra.sidecar.common.CQLSessionProvider;
+import org.apache.cassandra.sidecar.common.ClusterMembershipOperations;
+import org.apache.cassandra.sidecar.common.ICassandraAdapter;
+import org.apache.cassandra.sidecar.common.JmxClient;
+import org.apache.cassandra.sidecar.common.NodeSettings;
+import org.apache.cassandra.sidecar.common.StorageOperations;
+import org.apache.cassandra.sidecar.common.TableOperations;
+import org.apache.cassandra.sidecar.common.dns.DnsResolver;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A {@link ICassandraAdapter} implementation for Cassandra 4.0 and later
+ */
+public class CassandraAdapter implements ICassandraAdapter
```

Comment:
So this would make way more sense if we had left the 3.11 adapter in here - the idea is to track "anything from trunk backwards to 4.0" in the "trunk" package, and then if you have to split it some day (like the 3.11 stuff did) you'd create a new package and all of _those_ classes would have version-specific class names, so you'd never actually have a conflict. This way the default (and now only) classes don't need version-specific names because they handle significantly more than one version (4.0+ in this case), so it didn't make a lot of sense for it to be called `40`.

Reply:
There will be more version supported. The adapters should have the version in the name at the beginning, instead of a refactoring (guaranteed) in the future. The 3.11 example in fact supports that we should have version in the class name in the case of adapting to multiple versions, except Sidecar does not support 3.11 so we drop it. 

The concern in my original comment is not about conflicts. It is about clarity of the code, e.g. full qualified class name vs having version string in the simple class name, `org.apache.cassandra.sidecar.adapters.trunk.CassandraAdapter` vs. `Cassasndra50Adapter`.

Btw, I think it is not a good idea to use `trunk`. It is the "pointer" that can change to other versions. Sidecar should use 50 for clarity.

Reply:
recollecting how the versions work, I think we annotate a factory with `@MinimumVersion("....")`. So if the minimum version is `4.0.0` that means the adapter is good for any version 4.0.0+, although there might be breaking changes in new versions in which case a new `ICassandraFactory` must be provided.

The provided  factory in the code base  is good for `4.0.0`, `4.1.0`, and `5.0.0` looking at the test results. But the implementation we have was based on the minimum version of 4.0.0, so I think it would make more sense to have it be the `Cassandra40Factory`, which gives you a `Cassandra40Adapter`.

I think we should test against trunk for sure, this will allow us to detect any breaking changes and would let us determine whether Sidecar is _forward-compatible_ with the tip of Cassandra's trunk.

The question is what would be the expectation when latest is no longer compatible and we need to break it.

- latest remains compatible with the tip of Cassandra's trunk
- We introduce a Cassandra40Adapter then, that will work correctly with versions 4.0.0 through tip of trunk (not inclusive)

Reply:
After much discussion, we decided to leave this as-is (although renaming `latest` to `base` in the package name)